app/src/components/features/rules/RuleBuilder/Header/ActionButtons/CreateRuleButton/actions/index.js

```diff
@@ -3,10 +3,104 @@ import { actions } from "../../../../../../../../store";
+import {
+  extractInnerTextAndAttributesFromHTMLString,
+  getHTMLNodeName,
+  validateStringForSpecificHTMLNode,
+} from "./insertScriptValidators";
+import { isFeatureCompatible } from "utils/CompatibilityUtils";
+import FEATURES from "config/constants/sub/features";
+
+/**
+ * Convert user inputs into rule schema
+ *
+ * @param {*} ruleData
+ * @returns {
+ *  success: boolean,
+ *  newRuleData?: object, // in case of successful validation and transformation
+ *
+ *  // in case of failure
+ *  error?: string,
+ *  message?: string,
+ *  pairId?: string, // id of the pair that caused the error
+ * }
+ */
+export const transformAndValidateRuleFields = async (ruleData) => {
+  switch (ruleData.ruleType) {
+    case GLOBAL_CONSTANTS.RULE_TYPES.SCRIPT: {
+      const isCompatibleWithAttributesForScriptRule = isFeatureCompatible(FEATURES.SCRIPT_RULE_HTML_BLOCK);
+
+      if (!isCompatibleWithAttributesForScriptRule) {
+        return ruleData;
+      }
+
+      /* RuleData is explicitly read-only (reference to value in store) */
+      const ruleDataCopy = _.cloneDeep(ruleData);
+      const newPairs = [];
+
+      for (const pair of ruleDataCopy.pairs) {
+        const parsedPair = pair;
+        const newScripts = [];
+
+        for (const script of pair.scripts) {
+          let newScript = script;
+
+          const htmlNodeName = getHTMLNodeName(script.type, script.codeType);
+
+          if (script.type === GLOBAL_CONSTANTS.SCRIPT_TYPES.CODE) {
+            const res = await validateStringForSpecificHTMLNode(script.value, htmlNodeName);
+            if (!res.isValid) {
+              return {
+                error: res.validationError,
+                success: false,
+                message: res.errorMessage,
+                pairId: pair.id,
+              };
+            }
+            const { attributes, innerText: code } = extractInnerTextAndAttributesFromHTMLString(
+              script.value,
+              htmlNodeName
+            );
+            newScript = {
+              ...script,
+              attributes,
+              value: code,
+            };
+          } else if (script.type === GLOBAL_CONSTANTS.SCRIPT_TYPES.URL) {
+            const res = await validateStringForSpecificHTMLNode(script.wrapperElement, htmlNodeName);
+            if (!res.isValid) {
+              return {
+                error: res.validationError,
+                success: false,
+                message: res.errorMessage,
+                pairId: pair.id,
+              };
+            }
+            const { attributes } = extractInnerTextAndAttributesFromHTMLString(script.wrapperElement, htmlNodeName);
+            newScript = {
+              ...script,
+              attributes,
+            };
+          }
+          delete newScript["wrapperElement"]; // this was used to temporarily store the raw code
+          newScripts.push(newScript);
+        }
+        parsedPair.scripts = newScripts;
+        newPairs.push(parsedPair);
```

Comment:
One approach to reformat this code as

1. Instead of deep cloning the entire rule data, you just loop on the rulePairs
2. And just update the script object inside the rulePair upon transformation

It won't make code messy but should simplify the code. You will not have to use so many copies, like newScript, parsedRulePair etc.